### DISCUSSION TITLE
IPAM stack deployment

### DISCUSSION THREAD

**Discussion by vamsee33** [Azure/ipam#273] [Q&A] [answered]
When I am deploying with 2nd approach only services without identity , Should I include the UI and engine app IDs in the parameters.json file or directly in the PowerShell script without being expected from 1stdeployment? Also, any tips on setting up the SPN declaration without relying on the first deploymenti.e identies ?"

**Comment by DCMattyG** [comment]
Hi @vamsee33, I think I might need a bit more clarity on the issue you are having so I can provide a better answer.

If you're referring to the 2-part deployment, in step 1 the Azure Identities (App Registrations/Service Principals) are generated and their ID's and secrets are stored in a parameters file which is to be used in conjunction with step 2.

Step 2 then takes that parameters file and will deploy the Azure infrastructure/resources that will actually run the product.

More details can be found in our docs:
- Step 1: https://azure.github.io/ipam/#/deployment/README?id=azure-identities-only-deployment
- Step 2: https://azure.github.io/ipam/#/deployment/README?id=infrastructure-stack-only-deployment

I hope that helps, but if not please provide a few additional details as to where you're running into trouble and we'll be happy to assist however we can.

Thanks so much!

**Reply by vamsee33** [reply]
Hi @DCMattyG , thanks for your inputs. The Engine App SPN and UI App SPN were created manually and trying to do stack deployment itself. I'm using a powershell step in Github actions to deploy IPAM. Instead of passing client id and secrets from Engine And UI App spn from parameters file , Is there any way to pass the values from powershell script like using a flag ( example: -Engine_Client_ID 'value' Client_secret 'secret  fetching from kv with another step '?

Could you please let me know if need few more details and guide me

**Reply by vamsee33** [reply]
Like the below:


./deploy.ps1 `
  -Location "westus3" `
  -ParameterFile ./main.parameters.json `
  -NamePrefix "devipam
  -EngineAppID "steps.kvfetch.outputs.appid"
 -EngineClientSecret "steps.kvfetch.outputs.appsecret"

**Comment by DCMattyG** [comment, marked as answer]
Hi @vamsee33, I understand the ask much better now!

While there currently is not a way to pass those values in via command line, you can fill in the value in the `main.parameters.example.json` file located in the `deploy` folder, and then pass that into the Step 2 process by using the `-ParameterFile` switch as you were doing in the above example.

Thanks should get you where you need to be, but give it a try and let me know how it works.

**Reply by vamsee33** [reply]
Thanks @DCMattyG  for giving much clarification and responding promptly. One last concern regarding security as passing sensitive values as plain text in parameters.json file. Is there any way to pass the sensitive values (client id and secrets) I.e getting values from key vault and it can gettable using github actions??

**Reply by DCMattyG** [reply]
Hey @vamsee33, I think that whether you're reading from KeyVault and then passing it to the App Service, or putting it into a JSON file and passing it into the deployment script, one way or another it will briefly be plain text. That said, if this process is happening as a part of a GitHub action, that should be fairly secure and all of the communications with Azure via the PowerShell SDK should be SSL/TLS encrypted.

As far as reading a secret using a GitHub action, while I haven't done this personally I did a quick marketplace search for you and it does look as though there are some available actions to fetch secrets from KeyVault:

https://github.com/marketplace?query=keyvault

So for now, I would use these steps given your scenario:

1. Read in the `main.parameters.example.json` using something like `ConvertFrom-Json`
2. Fetch the ID's and secrets from the KeyVault
3. Update the value in the PowerShell object you created from the JSON file.
4. Output the object back to a JSON file
5. Execute the `deploy.ps1` script and pass in the newly created JSON file that contains the updated values

There could be other ways, but you might want to start there and see how that works.

Hope that helped!

**Reply by vamsee33** [reply]
Hi @DCMattyG ,

I followed the above approach which you stated
- name: Fetch secrets and update parameters
        shell: pwsh
        working-directory: deploy
        run: |
          $parameters = Get-Content -Path "main.parameters.example.json" | ConvertFrom-Json

          # Fetch secrets from Azure Key Vault
          $uiclientId = (Get-AzKeyVaultSecret -VaultName "lt-certdata" -Name "uiappid").SecretValueText
          $engineclientId = (Get-AzKeyVaultSecret -VaultName "lt-certdata" -Name "engineappid").SecretValueText
          $engineclientSecret = (Get-AzKeyVaultSecret -VaultName "lt-certdata" -Name "engineappsecret").SecretValueText

          # Update parameters.json object

          $parameters.parameters.uiAppId.value = $uiclientId
          $parameters.parameters.engineAppId.value = $engineclientId
          $parameters.parameters.engineAppSecret.value = $engineclientSecret

          # Save updated parameters to updated.parameters.json
          $parameters | ConvertTo-Json -Depth 10 | Set-Content -Path "main.parameters.json" 
          
      - name: Deploy Azure IPAM
        working-directory: deploy
        id: deployScript
        shell: pwsh
        run:  |  
              ./deploy.ps1 `
                -Location "westus3" `
                -ParameterFile ./main.parameters.json `
                

But was throwing an error
Contents of the latest error log file:
OperationStopped: /home/runner/work/Terraform-Azure/Terraform-Azure/deploy/deploy.ps1:801 Line |  801 |        throw [System.ArgumentException]::New("One of the required para …      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~      | One of the required parameters are missing or invalid.

As per detailed investigation failed at "Function Import-Parameters" in deploy.ps1
Could you please suggest anywhere to bypass Function Save-Parameters and Function Import-Parameters functions and can expect from the parameters file without creating dynamically with these two functions? and let us know skipping Identity deployment won't work out as SPN values are taking from 1st deployment and those are passing 2nd deployment
Appreciate if you would provide a necessary information!!

**Reply by DCMattyG** [reply]
@vamsee33, I'm pretty sure you aren't retrieving the secrets from KeyVault properly. Did you actually check the generated JSON file to confirm the appropriate secrets were being injected?

Here is the correct way to retrieve KeyVault secrets using the PowerShell SDK per our documentation:

https://learn.microsoft.com/en-us/azure/key-vault/secrets/quick-create-powershell#retrieve-a-secret-from-key-vault

```powershell
$secret = Get-AzKeyVaultSecret -VaultName "<your-unique-keyvault-name>" -Name "ExamplePassword" -AsPlainText
```

Can you give that a try and let me know if that helps?

**Reply by vamsee33** [reply]
@DCMattyG, Thanks for your quick find. It validates all the steps and started "New-AzSubscriptionDeployment" deployment but failed due to insufficient privileges. I will look into the prerequisites for service principal to deploy resources and let you know how it works
